### PR TITLE
Shut down agent when unit is not found

### PR DESCRIPTION
**Pull request by manadart**
This patch expedites the termination of the unit agent, when its unit is removed from state.

Previously, this condition was not surfaced as an error telling the agent to terminate, so there was potentially a restart before the condition was assessed resulting in shutdown.

The uniter now returns the same error as when the unit's `Life` indicates that it is `dead`.

## QA steps

- Deploy a unit, add a unit, remove a unit. 
- We should quiesce quite quickly without errors.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Shut down agent when unit not found](https://trello.com/c/DnL5iPCP/3360-shut-down-agent-when-unit-not-found)

File: worker/caasupgrader/upgrader.go [hunk 1]
```diff
@@ -178,7 +178,8 @@ func (u *Upgrader) loop() error {
 		logger.Debugf("%s requested for %v from %v to %v", direction, u.tag, jujuversion.Current, wantVersion)
 		err = u.operatorUpgrader.Upgrade(u.tag.String(), wantVersion)
 		if err != nil {
-			return errors.Annotatef(err, "requesting upgrade for %v from %v to %v", u.tag, jujuversion.Current, wantVersion)
+			return errors.Annotatef(
+				err, "requesting upgrade for %v from %v to %v", u.tag.String(), jujuversion.Current, wantVersion)
```
Comment (manadart) [inline worker/caasupgrader/upgrader.go:182]: This is just a drive-by to fix the `!s{...}` spam in the log line.

**Comment by manadart** [comment]
!!build!!

File: worker/uniter/uniter.go [hunk 2]
```diff
@@ -672,21 +672,27 @@ func (u *Uniter) init(unitTag names.UnitTag) (err error) {
...
 	if u.unit.Life() == life.Dead {
-		// If we started up already dead, we should not progress further. If we
-		// become Dead immediately after starting up, we may well complete any
-		// operations in progress before detecting it; but that race is fundamental
-		// and inescapable, whereas this one is not.
+		// If we started up already dead, we should not progress further.
+		// If we become Dead immediately after starting up, we may well
+		// complete any operations in progress before detecting it,
+		// but that race is fundamental and inescapable,
+		// whereas this one is not.
```
Comment (wallyworld) [inline worker/uniter/uniter.go:688]: This comment is applicable to the not found case also I think.
Would it be nice to do this?
(also avoids u.stopUnitError() twice)

```
	u.unit, err = u.st.Unit(unitTag)
	if err != nil && !errors.IsNotFound(err) {
		return errors.Trace(err)
    }
	if err != nil || u.unit.Life() == life.Dead {
		// If we started up already dead, we should not progress further. If we
		// If we started up already dead, we should not progress further.
		// become Dead immediately after starting up, we may well complete any
		// If we become Dead immediately after starting up, we may well
		// operations in progress before detecting it; but that race is fundamental
		// complete any operations in progress before detecting it,
		// and inescapable, whereas this one is not.
		// but that race is fundamental and inescapable,
		// whereas this one is not.
        return u.stopUnitError()
    }
```

**Review by wallyworld — APPROVED** [review]

File: worker/uniter/uniter.go (hunk 2 above)
Comment (manadart) [reply worker/uniter/uniter.go:688]: I've moved the comment higher, but I think assessing the error in a single block reads better.

**Comment by manadart** [comment]
$$merge$$
